codec/src/main/java/io/netty/handler/codec/ByteString.java

```diff
@@ -0,0 +1,769 @@
+     *
+     * @return {@code -1} if the processor iterated to or beyond the end of the specified area.
+     *         The last-visited index If the {@link ByteStringProcessor#process(byte)} returned {@code false}.
+     */
+    public int forEachByte(int index, int length, ByteStringProcessor processor) throws Exception {
+        if (index < 0 || length > value.length - index) {
+            throw new IndexOutOfBoundsException("expected: " + "0 <= index(" + index + ") <= start + length(" + length
+                    + ") <= " + "length(" + value.length + ')');
+        }
+
+        for (int i = index; i < length; ++i) {
+            if (!processor.process(value[i])) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    /**
+     * Iterates over the readable bytes of this buffer with the specified {@code processor} in descending order.
+     *
+     * @return {@code -1} if the processor iterated to or beyond the beginning of the readable bytes.
+     *         The last-visited index If the {@link ByteBufProcessor#process(byte)} returned {@code false}.
+     */
+    public int forEachByteDesc(ByteStringProcessor processor) throws Exception {
+        return forEachByteDesc(0, length(), processor);
+    }
+
+    /**
+     * Iterates over the specified area of this buffer with the specified {@code processor} in descending order.
+     * (i.e. {@code (index + length - 1)}, {@code (index + length - 2)}, ... {@code index}).
+     *
+     * @return {@code -1} if the processor iterated to or beyond the beginning of the specified area.
+     *         The last-visited index If the {@link ByteBufProcessor#process(byte)} returned {@code false}.
+     */
+    public int forEachByteDesc(int index, int length, ByteStringProcessor processor) throws Exception {
+        if (index < 0 || length > value.length - index) {
+            throw new IndexOutOfBoundsException("expected: " + "0 <= index(" + index + ") <= start + length(" + length
+                    + ") <= " + "length(" + value.length + ')');
+        }
+
+        for (int i = index + length - 1; i >= index; --i) {
+            if (!processor.process(value[i])) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    public byte byteAt(int index) {
+        return value[index];
+    }
+
+    public boolean isEmpty() {
+        return value.length == 0;
+    }
+
+    public int length() {
+        return value.length;
+    }
+
+    /**
+     * During normal use cases the {@link ByteString} should be immutable, but if the underlying array is shared,
+     * and changes then this needs to be called.
+     */
+    public void arrayChanged() {
+        hash = 0;
+    }
+
+    /**
+     * This gives direct access to the underlying storage array.
+     * The {@link #toByteArray()} should be preferred over this method.
+     * If the return value is changed then {@link #arrayChanged()} must be called.
+     */
+    public byte[] array() {
```

Comment:
Note that this method is not ideal...it would be nice not to need it but we use it in HTTP/2 for the HPACK interface. Once we restructure twitter's HPACK we may not need it...but still may be useful for others...?